kernel/libs/aster-bigtcp/src/iface/common.rs

```diff
@@ -122,58 +123,78 @@ impl<E: Ext> IfaceCommon<E> {
         iface: Arc<dyn Iface<E>>,
         config: BindPortConfig,
     ) -> core::result::Result<BoundPort<E>, BindError> {
-        let port = self.bind_port(config)?;
-        Ok(BoundPort { iface, port })
+        let (port, port_state) = self.bind_port(config)?;
+        Ok(BoundPort {
+            iface,
+            port,
+            port_state,
+        })
     }
 
     /// Allocates an unused ephemeral port.
     ///
     /// We follow the port range that many Linux kernels use by default, which is 32768-60999.
     ///
     /// See <https://en.wikipedia.org/wiki/Ephemeral_port>.
-    fn alloc_ephemeral_port(&self) -> Option<u16> {
+    fn alloc_ephemeral_port(&self, can_reuse: bool) -> Option<u16> {
         let mut used_ports = self.used_ports.lock();
         for port in IP_LOCAL_PORT_START..=IP_LOCAL_PORT_END {
             if let Entry::Vacant(e) = used_ports.entry(port) {
-                e.insert(0);
+                e.insert(Arc::new(PortState::new(0, can_reuse)));
                 return Some(port);
             }
         }
         None
     }
 
-    fn bind_port(&self, config: BindPortConfig) -> Result<u16, BindError> {
+    fn bind_port(&self, config: BindPortConfig) -> Result<(u16, Arc<PortState>), BindError> {
         let port = if let Some(port) = config.port() {
             port
         } else {
-            match self.alloc_ephemeral_port() {
+            let can_reuse = config.can_reuse();
+            match self.alloc_ephemeral_port(can_reuse) {
                 Some(port) => port,
-                None => return Err(BindError::Exhausted),
+                None => {
+                    // FIXME: If `config.can_reuse` is `true`,
+                    // we should check all ports currently in use to determine if any can be reused,
+                    // rather than returning an error immediately.
+                    return Err(BindError::Exhausted);
+                }
             }
         };
 
         let mut used_ports = self.used_ports.lock();
 
-        if let Some(used_times) = used_ports.get_mut(&port) {
-            if *used_times == 0 || config.can_reuse() {
-                // FIXME: Check if the previous socket was bound with SO_REUSEADDR.
-                *used_times += 1;
+        let port_state = if let Some(port_state) = used_ports.get(&port) {
+            // FIXME: If the socket is not a backlog socket,
+            // we should check whether there is a listening socket on the port.
+            // If there is, the socket cannot be bound to that port.
+            let can_reuse = matches!(config, BindPortConfig::Backlog(_))
+                || (port_state.can_reuse() & config.can_reuse());
+            let nsocket = port_state.nsocket.load(Ordering::Relaxed);
+            if nsocket == 0 || can_reuse {
```

Comment:
This is racy. I would prefer not to use `Arc` here. We can have:
```rust
    used_ports: SpinLock<BTreeMap<u16, PortState>, BottomHalfDisabled>,
```
```rust
pub struct PortState {
    nsocket: usize,
    can_reuse: bool,
}
```

Therefore, we can properly utilize the `used_ports` spin lock to prevent unnecessary race conditions. This creates the natural invariant that `port_state.can_reuse || port_state.nsocket == 1` is always true.